From b5bd07356492ef9736eab70f27028db0fb8b3516 Mon Sep 17 00:00:00 2001
From: David Bauer <mail@david-bauer.net>
Date: Sat, 17 Jan 2026 23:12:47 +0100
Subject: [PATCH 3/3] wifi: mt76: mt7915: sync power save state with WA

Host receives a power save state change event everytime a station enters
or leaves PS mode. Use this unsolicited event to stop filling up TX
queues in firmware, as the hardware will buffer frames for stations in
power safe indefinitely.

Without this change, one can see AQL starting to misbehave due to
excessive queueing of frames in hardware. This is especially worse if
station enter power save and never leave this state while continuing to
receive data.

Signed-off-by: David Bauer <mail@david-bauer.net>
---
 ...mt7915-sync-power-save-state-with-WA.patch | 114 ++++++++++++++++++
 1 file changed, 114 insertions(+)
 create mode 100644 patches/openwrt/0014-wifi-mt76-mt7915-sync-power-save-state-with-WA.patch

diff --git a/patches/openwrt/0014-wifi-mt76-mt7915-sync-power-save-state-with-WA.patch b/patches/openwrt/0014-wifi-mt76-mt7915-sync-power-save-state-with-WA.patch
new file mode 100644
index 00000000..90035805
--- /dev/null
+++ b/patches/openwrt/0014-wifi-mt76-mt7915-sync-power-save-state-with-WA.patch
@@ -0,0 +1,114 @@
+From: David Bauer <mail@david-bauer.net>
+Date: Sat, 17 Jan 2026 23:11:43 +0100
+Subject: wifi: mt76: mt7915: sync power save state with WA
+
+Host receives a power save state change event everytime a station enters
+or leaves PS mode. Use this unsolicited event to stop filling up TX
+queues in firmware, as the hardware will buffer frames for stations in
+power safe indefinitely.
+
+Without this change, one can see AQL starting to misbehave due to
+excessive queueing of frames in hardware. This is especially worse if
+station enter power save and never leave this state while continuing to
+receive data.
+
+Signed-off-by: David Bauer <mail@david-bauer.net>
+
+diff --git a/package/kernel/mt76/patches/401-wifi-mt76-mt7915-sync-power-save-state-with-WA.patch b/package/kernel/mt76/patches/401-wifi-mt76-mt7915-sync-power-save-state-with-WA.patch
+new file mode 100644
+index 0000000000000000000000000000000000000000..6fe725dcaf0438818e32bfa7d01465fb0ec823ad
+--- /dev/null
++++ b/package/kernel/mt76/patches/401-wifi-mt76-mt7915-sync-power-save-state-with-WA.patch
+@@ -0,0 +1,92 @@
++From 4b282c1c8529559130d127fe022fdaed4452fa67 Mon Sep 17 00:00:00 2001
++From: David Bauer <mail@david-bauer.net>
++Date: Fri, 16 Jan 2026 21:53:22 +0100
++Subject: [PATCH 2/2] wifi: mt76: mt7915: sync power save state with WA
++
++Host receives a power save state change event everytime a station enters
++or leaves PS mode. Use this unsolicited event to stop filling up TX
++queues in firmware, as the hardware will buffer frames for stations in
++power safe indefinitely.
++
++Without this change, one can see AQL starting to misbehave due to
++excessive queueing of frames in hardware. This is especially worse if
++station enter power save and never leave this state while continuing to
++receive data.
++
++Signed-off-by: David Bauer <mail@david-bauer.net>
++---
++ mt7915/mcu.c | 30 ++++++++++++++++++++++++++++++
++ mt7915/mcu.h |  9 +++++++++
++ 2 files changed, 39 insertions(+)
++
++diff --git a/mt7915/mcu.c b/mt7915/mcu.c
++index 40364eea..59697cc9 100644
++--- a/mt7915/mcu.c
+++++ b/mt7915/mcu.c
++@@ -402,6 +402,33 @@ mt7915_mcu_rx_bcc_notify(struct mt7915_dev *dev, struct sk_buff *skb)
++ 			mt7915_mcu_cca_finish, mphy->hw);
++ }
++ 
+++static void mt7915_mcu_rx_ps_sync(struct mt7915_dev *dev, struct sk_buff *skb)
+++{
+++	struct mt7915_mcu_ps_notify *p;
+++	struct ieee80211_sta *sta;
+++	struct mt76_wcid *wcid;
+++	u16 wcid_idx;
+++
+++	p = (struct mt7915_mcu_ps_notify *)skb->data;
+++
+++	wcid_idx = p->wtbl_lower | (p->wtbl_higher) << 8;
+++	wcid = mt76_wcid_ptr(dev, wcid_idx);
+++
+++	if (!wcid)
+++		return;
+++
+++	sta = wcid_to_sta(wcid);
+++	if (!sta)
+++		return;
+++
+++	if (p->ps_bit)
+++		set_bit(MT_WCID_FLAG_PS, &wcid->flags);
+++	else
+++		clear_bit(MT_WCID_FLAG_PS, &wcid->flags);
+++
+++	ieee80211_sta_ps_transition(sta, !!p->ps_bit);
+++}
+++
++ static void
++ mt7915_mcu_rx_ext_event(struct mt7915_dev *dev, struct sk_buff *skb)
++ {
++@@ -424,6 +451,9 @@ mt7915_mcu_rx_ext_event(struct mt7915_dev *dev, struct sk_buff *skb)
++ 	case MCU_EXT_EVENT_BCC_NOTIFY:
++ 		mt7915_mcu_rx_bcc_notify(dev, skb);
++ 		break;
+++	case MCU_EXT_EVENT_PS_SYNC:
+++		mt7915_mcu_rx_ps_sync(dev, skb);
+++		break;
++ 	default:
++ 		break;
++ 	}
++diff --git a/mt7915/mcu.h b/mt7915/mcu.h
++index 3be105bc..7e6b12b3 100644
++--- a/mt7915/mcu.h
+++++ b/mt7915/mcu.h
++@@ -54,6 +54,15 @@ struct mt7915_mcu_bcc_notify {
++ 	u8 rsv;
++ } __packed;
++ 
+++struct mt7915_mcu_ps_notify {
+++	struct mt76_connac2_mcu_rxd_hdr rxd;
+++
+++	u8 wtbl_lower;
+++	u8 ps_bit;
+++	u8 wtbl_higher;
+++	u8 rsv;
+++} __packed;
+++
++ struct mt7915_mcu_rdd_report {
++ 	struct mt76_connac2_mcu_rxd_hdr rxd;
++ 
++-- 
++2.51.0
++
-- 
2.51.0

